src/rgw/rgw_rest_s3.cc

```diff
@@ -6466,10 +6473,19 @@ int RGWSelectObj_ObjStore_S3::range_request(int64_t ofs, int64_t len,void *buff,
   requested_buffer.clear();
   m_request_range = len;
 
-  ldout(s->cct, 10) << "S3select: calling execute(async):" << " offset :" << ofs << " length :" << len << "buffer size : " << requested_buffer.size() << dendl;
+  ldout(s->cct, 10) << "S3select: calling execute(async):" << " request-offset :" << ofs << " request-length :" << len << " buffer size : " << requested_buffer.size() << dendl;
 
   range_request_complete = std::promise<void>();
-  range_request_complete_future = range_request_complete.get_future();
+  
+  try {
+    range_request_complete_future = range_request_complete.get_future();
```

Comment:
what's the motivation for promise/future here? is something running in a different thread?

Reply:
i see that `RGWSelectObj_ObjStore_S3::execute()` calls `range_request()`, which calls `RGWGetObj::execute()`. that will generate one or more calls to `RGWSelectObj_ObjStore_S3::send_response_data()`

so you're waiting for `send_response_data()` to get everything it needed? but since we're calling `RGWGetObj::execute()` synchronously, doesn't the fact that it returned successfully already guarantee that it fed everything to `send_response_data()`?

Reply:
the `send_response_data()` is a callback.
been called a-sync. (upon calling to  `RGWGetObj::execute()` )

and yes, in some cases it takes more than one call (of `send_response_data()`).

does `RGWGetObj::execute()` is a sync function for any case? (always returns upon request completion)

Reply:
GetObj does not keep the buffers in memory after passing them to `send_response_data()`. it should be reading chunks of up to 4MB (rgw_max_chunk_size) from rados, and may stream up to 4 of those requests in parallel (controlled by rgw_get_obj_window_size=16MB)

so i'd expect a given GetObj request to have at most 16MB of buffers allocated at a time